test/projection.jl

```diff
@@ -1,240 +1,285 @@
+
+        @test ProjectTo(Any[1, 2])(1:2) == [1.0, 2.0]  # projects each number.
+        @test Tuple(ProjectTo(Any[1, 2+3im])(1:2)) === (1.0, 2.0 + 0.0im)
+        @test ProjectTo(Any[true, false]) isa ProjectTo{NoTangent}
+
+        # empty arrays
+        @test isempty(ProjectTo([])(1:0))
+        @test_throws DimensionMismatch ProjectTo(Int[])([2])
+        @test ProjectTo(Bool[]) isa ProjectTo{NoTangent}
     end
 
-    @testset "to Real" begin
-        # Float64
-        @test 3.2 == ProjectTo(1.0)(3.2)
-        @test 0.0 == ProjectTo(1.0)(ZeroTangent())
-        @test 3.2 == ProjectTo(1.0)(@thunk(3.2))
+    @testset "Base: zero-arrays & Ref" begin
+        pzed = ProjectTo(fill(1.0))
+        @test pzed(fill(3.14)) == fill(3.14)  # easy
+        @test pzed(fill(3)) == fill(3.0)      # broadcast type change must not produce number
+        @test pzed(hcat(3.14)) == fill(3.14)  # reshape of length-1 array
+        @test pzed(3.14 + im) == fill(3.14)   # re-wrap of a scalar number
+
+        @test_throws DimensionMismatch ProjectTo([1])(3.14 + im) # other array projectors don't accept numbers
+        @test_throws DimensionMismatch ProjectTo(hcat([1,2]))(3.14)
+        @test pzed isa ProjectTo{AbstractArray}
+
+        pref = ProjectTo(Ref(2.0))
+        @test pref(Ref(3+im))[] === 3.0
+        @test pref(4)[] === 4.0  # also re-wraps scalars
+        @test pref(Ref{Any}(5.0)) isa Base.RefValue{Float64}
+        pref2 = ProjectTo(Ref{Any}(6+7im))
+        @test pref2(Ref(8))[] === 8.0 + 0.0im
+
+        prefvec = ProjectTo(Ref([1,2,3+4im]))  # recurses into contents
+        @test prefvec(Ref(1:3)) isa Base.RefValue{Vector{ComplexF64}}
+        @test_throws DimensionMismatch prefvec(Ref{Any}(1:5))
+    end
 
-        # down
-        @test 3.2 == ProjectTo(1.0)(3.2 + 3im)
-        @test 3.2f0 == ProjectTo(1.0f0)(3.2)
-        @test 3.2f0 == ProjectTo(1.0f0)(3.2 - 3im)
+    @testset "LinearAlgebra: $adj vectors" for adj in [transpose, adjoint]
+        # adjoint vectors
+        padj = ProjectTo(adj([1,2,3]))
+        adjT = typeof(adj([1,2,3.0]))
+        @test padj(transpose(1:3)) isa adjT
+        @test padj([4 5 6+7im]) isa adjT
+        @test padj([4.0 5.0 6.0]) isa adjT
+
+        @test_throws DimensionMismatch padj([1,2,3])
+        @test_throws DimensionMismatch padj([1 2 3]')
+        @test_throws DimensionMismatch padj([1 2 3 4])
+
+        padj_complex = ProjectTo(adj([1,2,3+4im]))
+        @test padj_complex([4 5 6+7im]) == [4 5 6+7im]
+        @test padj_complex(transpose([4, 5, 6+7im])) == [4 5 6+7im]
+        @test padj_complex(adjoint([4, 5, 6+7im])) == [4 5 6-7im]
+
+        # evil test case
+        xs = adj(Any[Any[1,2,3], Any[4+im,5-im,6+im,7-im]])
+        pvecvec3 = ProjectTo(xs)
+        @test pvecvec3(xs)[1] == [1 2 3]
+        @test pvecvec3(xs)[2] == adj.([4+im 5-im 6+im 7-im])
+        @test pvecvec3(xs)[2] isa LinearAlgebra.AdjOrTransAbsMat{ComplexF64, <:Vector}
```

Comment:
I thought I finally got this working, recursive transpose is a pain. But now I get weird errors on 1.5 & 1.6 CI:
```
566
  Expression: pvecvec3(ys) isa LinearAlgebra.AdjOrTransAbsVec
567
  Element type mismatch. Tried to create an `Adjoint{Adjoint}` from an object with eltype `Vector{T} where T`, but the element type of the adjoint of an object with eltype `Vector{T} where T` must be `Adjoint{_A, _B} where {_A, _B}`.
568
  Stacktrace:
569
    [1] error(s::String)
570
      @ Base ./error.jl:33
571
    [2] checkeltype_adjoint(#unused#::Type{Adjoint}, #unused#::Type{Vector{T} where T})
572
      @ LinearAlgebra /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/LinearAlgebra/src/adjtrans.jl:76
573
    [3] Adjoint{Adjoint, Vector{Vector{T} where T}}(A::Vector{Vector{T} where T})
574
      @ LinearAlgebra /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/LinearAlgebra/src/adjtrans.jl:38
575
    [4] Adjoint(A::Vector{Vector{T} where T})
576
      @ LinearAlgebra /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/LinearAlgebra/src/adjtrans.jl:95
577
    [5] adjoint(A::Vector{Vector{T} where T})
578
      @ LinearAlgebra /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/LinearAlgebra/src/adjtrans.jl:136
579
    [6] (::ProjectTo{Adjoint, NamedTuple{(:parent,), Tuple{ProjectTo{AbstractArray, NamedTuple{(:elements, :axes), Tuple{Vector{ProjectTo{AbstractArray, D} where D<:NamedTuple}, Tuple{Base.OneTo{Int64}}}}}}}})(dx::Matrix{Matrix{Any}})
    ```